### DISCUSSION TITLE
Some Copy-mode ideas

### DISCUSSION THREAD

**Discussion by linsomniac** [wezterm/wezterm#2547] [Ideas]
Some of these may be things I just haven't figured out yet, please let me know if that is the case.  Some of them are probably things that I can implement in a custom keybinding, and I'll probably play with that as I get time, but they might also be useful to have upstream.  Some of them maybe y'all will just have better ideas than I do.

1. I'm loving "y" in copy mode!  What about "Y" to match Ctrl-Shift-Space capital mode to copy and immediately paste?
2. What are the default patterns for Ctrl-Shift-Space, I searched through the code and docs yesterday and couldn't figure it out, the docs only gave an example of matching SHAs.  I'm thinking about augmenting the list to match my needs more closely.
3. This isn't fully formed: I often feel like I want to widen a match.  Maybe a mode where you select the pattern with letters, but then are in kind of a "v" mode with that match highlighted and you can then modify the match, with some quick keys for "select big word around match"?
4. Maybe in Ctrl-Shift-Space mode have some "switch pattern" mappings to flip the match from the default to specialized ones like: "words", "big words", "IP addresses", "git hashes", "commands" (after prompt to end of line), "lines"?
5. A few more Ctrl-Shift-X mappings: W (next "big" word, basically whitespace delimited), B (previous "big" word), I find myself wanting to use "e" and "E" to move to the end of the current word, but that may be a bit too specialized.

I'm really loving Copy-mode, use it all the time!

**Comment by wez** [comment]
> 1\. I'm loving "y" in copy mode!  What about "Y" to match Ctrl-Shift-Space capital mode to copy and immediately paste?

You can add an assignment for that:

```lua
local wezterm = require 'wezterm'
local act = wezterm.action

local copy_mode = nil
if wezterm.gui then
  copy_mode = wezterm.gui.default_key_tables().copy_mode

  -- Depending on the platform, uppercase assignments
  -- may or may not have SHIFT listed as the modifier,
  -- so register with both just to be sure.
  for _, mods in ipairs{'NONE', 'SHIFT'} do
    table.insert(
      copy_mode,
      {
        key = 'Y',
        mods = mods,
        action = act.Multiple {
          { CopyTo = 'ClipboardAndPrimarySelection' },
          { CopyMode = 'Close' },
          { PasteFrom = 'PrimarySelection' },
        },
      }
    )
  end
end

return {
  key_tables = {
    copy_mode = copy_mode
  }
}
```

> 2\. What are the default patterns for Ctrl-Shift-Space, I searched through the code and docs yesterday and couldn't figure it out, the docs only gave an example of matching SHAs.  I'm thinking about augmenting the list to match my needs more closely.

You can find them here:
https://github.com/wez/wezterm/blob/main/wezterm-gui/src/overlay/quickselect.rs#L26-L55


> 3\. This isn't fully formed: I often feel like I want to widen a match.  Maybe a mode where you select the pattern with letters, but then are in kind of a "v" mode with that match highlighted and you can then modify the match, with some quick keys for "select big word around match"?

You can add a key assignment that activates word selection mode.
Take a look at https://wezfurlong.org/wezterm/config/lua/keyassignment/CopyMode/SetSelectionMode.html

> 4\. Maybe in Ctrl-Shift-Space mode have some "switch pattern" mappings to flip the match from the default to specialized ones like: "words", "big words", "IP addresses", "git hashes", "commands" (after prompt to end of line), "lines"?

Interesting idea.  Not possible today, but you could define a key binding for each of those using [QuickSelectArgs](https://wezfurlong.org/wezterm/config/lua/keyassignment/QuickSelectArgs.html)

> 5\. A few more Ctrl-Shift-X mappings: W (next "big" word, basically whitespace delimited), B (previous "big" word), I find myself wanting to use "e" and "E" to move to the end of the current word, but that may be a bit too specialized.

In the nightly you can use `f<SPACE>` to jump forwards to the next space on the line, then use `;` to jump to the next occurrence in the line, or `,` to jump backwards.
You can use `t<SPACE>` to explicitly jump backwards.  `<SPACE>` can be any character, so you could `f,` to jump to a comma and so on.

`F` and `T` are the same but jump to the character preceding the match.

See: https://github.com/wez/wezterm/issues/2528

**Reply by linsomniac** [reply]
Re f/t/F/T: have you considered switching those assignments to be:

- fX: Move forward to X
- FX: Move backwards to X
- tX: Move forward to just before X
- TX: Move backwards to just before X

This is the "vi" mapping, and the mapping you describe above is similar but different enough to be confusing.

Thanks for adding "f", that's something that I've been missing!  I'm using copy mode a lot.

**Comment by linsomniac** [comment]
Thank you for that "Y" snippet.  I'm getting an error while trying to include that into my existing config:

    07:34:23.965  ERROR  mux::connui             > Configuration Error: runtime error: [string "/Users/sean/.config/wezterm/wezterm.lua"]:6: attempt to call a nil value (field 'default_key_tables')
    stack traceback:
	    [string "/Users/sean/.config/wezterm/wezterm.lua"]:6: in main chunk
    07:34:23.981  ERROR  wezterm_gui::termwindow > Failed to apply config overrides to window: runtime error: [string "/Users/sean/.config/wezterm/wezterm.lua"]:6: attempt to call a nil value (field 'default_key_tables')
    stack traceback:
	    [string "/Users/sean/.config/wezterm/wezterm.lua"]:6: in main chunk: nil

Where line 6 is `copy_mode = wezterm.gui.default_key_tables().copy_mode`.  I think this is saying that default_key_tables is nil when it's trying to call it?  I'm running the 09/05 wezterm.

**Reply by wez** [reply]
You need the nightly build for that function: https://wezfurlong.org/wezterm/config/lua/wezterm.gui/default_key_tables.html

**Comment by linsomniac** [comment]
Finally had a chance to grab the nightly and just to clarify the f/F/t/T work as I expect, I must just misunderstand your description of them.  Can't wait to try out that "Y" map, thanks so much!

**Comment by pengux** [comment]
How about merging the search overlay with copy mode? It's unfortunate that you cannot search, then using copy mode navigation to move around and adjust the selection for example.

**Reply by pengux** [reply]
I've realized you could achieve it with a few key mappings but it doesn't really work satisfactory. For example:

1. Enter copy mode
2. Hit "/" to start searching
3. Hit "Enter" to go to prior match until you are at the match that you'd want
4. Hit "w" to select the word (or any other navigation keys in copy mode)
5. Hit "y" to yank and close copy mode

**Reply by GP4cK** [reply]
The search doesn't work for me... Would love to have something like [flash.nvim](https://github.com/folke/flash.nvim) where you type a few letters and the app generates a letter to quickly jump to the location (similar to quick select)
